File: leehyowonzero/12week/72413.py
import sys
INF = sys.maxsize

def floyd(dp, n):
    for k in range(n+1):
        for i in range(n+1):
            for j in range(n+1):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j])
    return dp
    


def solution(n, s, a, b, fares):
    answer = INF
    
    dp = [[INF for _ in range(n+1)] for _ in range(n+1)]
    for i in range(n+1):
        dp[i][i] = 0
    for fare in fares:
        c, d, f = fare
        dp[c][d] = f
        dp[d][c] = f
    
    dp = floyd(dp, n)
    
    for k in range(1, n+1):
        answer = min(answer, dp[s][k] + dp[k][a] + dp[k][b])
        
    return answer

File: JeongGod/12week/programmers/72413.java
import java.util.*;
class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = Integer.MAX_VALUE;
        int[][] graph = new int[n][n];
        
        for (int i=0; i<n; i++) {
            Arrays.fill(graph[i], 1000000);
            
            graph[i][i] = 0;
        }
        for (int[] x : fares) {
            graph[x[0]-1][x[1]-1] = x[2];
            graph[x[1]-1][x[0]-1] = x[2];
        }
        
        for (int k=0; k<n; k++) {
            for (int i=0; i<n; i++) {
                for (int j=0; j<n; j++) {
                    graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
                }
            }
        }

        
        for (int i=0; i<n; i++) {
            answer = Math.min(answer, graph[s-1][i] + graph[i][a-1] + graph[i][b-1]);
        }
        
        return answer;
    }
}